File: ml-benchmark/synduce/constraints/memo/mts_memo_sum.f
Inductive List = elt Int | cons {Int, List};
Inductive MList = ielt Int | icons {Int, Int, MList};

repr = fix (
  \f: MList -> List. \m: MList.
  match m with
    ielt a -> elt a
  | icons {h, _, t} -> cons {h, f t}
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt a -> a
  | cons {h, t} -> + h (f t)
  end
);

is_memo = fix (
  \f: MList -> Bool. \m: MList.
  match m with
    ielt _ -> true
  | icons {h, s, t} -> and (== s (sum (repr m))) (f t)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> max (f t) (sum xs)
  end
);

target = fix (
  \f: MList -> Compress MList. \m: MList.
  match m with
    ielt x -> m
  | icons {h, s, t} -> icons {h, s, f t}
  end
);

/* Customized generator, not used
gen = fix (
  \f: Tree -> TreeMemo. \t: Tree.
  match t with
    leaf a -> mleaf {if < a 2 then 1 else 0, a}
  | node {a, l, r} ->
    let res = {f l, f r} in
      mnode {+ (if < a 2 then 1 else 0) (+ (memo res.1) (memo res.2)), a, res.1, res.2}
  end
);*/

main = /*\t: Tree. let mt = gen t in*/
  \m: MList.
    if is_memo m then spec (repr (target m)) else 0;

File: ml-benchmark/synduce/constraints/memo/mts_memo_sum.ml
type tlist = Elt of int | Cons of int * tlist
type mlist = Ielt of int | Icons of int * int * mlist

let rec repr m =
  match m with
  | Ielt a -> Elt a
  | Icons (h, _, t) ->
      let r = repr t in
      Cons (h, r)

let rec sum xs =
  match xs with
  | Elt a -> a
  | Cons (h, t) ->
      let st = sum t in
      h + st

let rec is_memo m =
  match m with
  | Ielt _ -> true
  | Icons (_, s, t) ->
      let eq =
        let r = repr m in
        let v = sum r in
        s == v
      in
      let recres = is_memo t in
      eq && recres

let max a b = if a < b then b else a

let rec spec xs =
  match xs with
  | Elt x -> x
  | Cons (_, t) ->
      let ft = spec t in
      let su = sum xs in
      max ft su

val target: mlist -> mlist compress
let rec target m =
  match m with
  | Ielt _ -> m
  | Icons (h, s, t) ->
      let ft = target t in
      Icons (h, s, ft)

let program m =
  if is_memo m then
    let r = target m in
    let rr = repr r in
    spec rr
  else
    0